Allow parsing Content-Type and Accept headers with version

pgomulka opened the pull request:
Content-Type and Accept headers expect a versioned form of media types
like application/vnd.elasticsearch+json;compatible-with=7
when previously it was simple application/json or (cbor, yaml..) - it is still
supported


next step after https://github.com/elastic/elasticsearch/pull/61987
relates https://github.com/elastic/elasticsearch/pull/60516

pgomulka committed:
Allow parsing Content-Type and Accept headers with version

Content-Type and Accept headers expect a verioned form of media types
like application/vnd.elasticsearch+json;compatible-with=7
when previously it was simple application/json or similar - it is still
supported

jakelandis commented on server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java:
```diff
@@ -84,4 +84,57 @@ public void testFromRubbish() throws Exception {
         assertThat(XContentType.fromMediaTypeOrFormat("gobbly;goop"), nullValue());
     }
+
+    public void testMediaType() {
```
can you add a test for yaml too ?

pgomulka replied:
will do

sethmlarson commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -114,13 +116,19 @@ public XContent xContent() {
     };
 
+    private static final Pattern COMPATIBLE_API_HEADER_PATTERN = Pattern.compile(
+        "(application|text)/(vnd.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?",
```
Regex needs a few more escapes:
- Escape the `.` within `vnd.elasticsearch+`
- Escape the `/` between the type and subtype

 `(application|text)\\/(vnd\\.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?`

dakrone replied:
Something I am curious about is the `(\\d+))?` at the end. If possible, it would be nice to define this as part of our "spec". This would be something we could be strict (eschew leniency!) about, such as enforcing it to be in a specific format (and throwing an exception when someone specifies a pattern that is not correct.

If you agree with the sentiment, I would be curious what we expect the final number to be, for example, which of these do we intend to support (now *and* in the future):

- compatible-with=7
- compatible-with=7.11
- compatible-with=7.x
- compatible-with=7.11.2

Then I think we should make this strict so users find out early if they've send an illegal value

pugnascotia replied:
@sethmlarson Why does the forward slash needs escaping? It's not special in Java?

I wondering if we should anchor the pattern with `^` at the start and `$` at the end, to be on the safe side?

In `([^;]+)(\\s*;`, the `[^;]+` will also consume whitespace, so I believe you can remove the `\\s*`.

Can we switch any of these capture groups to non-capturing if we don't actually need the contents, just the grouping? i.e `(?:<pattern>)`

sethmlarson replied:
@pugnascotia You're right the `/` doesn't need an escape in Java

rjernst commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -114,13 +116,19 @@ public XContent xContent() {
     };
 
+    private static final Pattern COMPATIBLE_API_HEADER_PATTERN = Pattern.compile(
+        "(application|text)/(vnd.elasticsearch\\+)?([^;]+)(\\s*;\\s*compatible-with=(\\d+))?",
```
> (vnd.elasticsearch\\+)?

Am I reading this correctly that the `vnd.elasticsearch` part is completely optional? I thought the intent was to only add `compatible-with` when `vnd.elasticsearch` is present?

pgomulka replied:
you are right. This is possibly a bit oversimplified.
Do you think we should aim to be more strict on the regex side or should we validate the values after the parsing with more relaxed regex?

The more strict could look like 
```
Pattern.compile(
        "(application|text)/((vnd\\.elasticsearch\\+(json|smile|yaml|cbor)(\\s*;\\s*compatible-with=(\\d+)))" +
            "|(json|smile|yaml|cbor))");
```

there is a simplistic validation in `parseVerion` method, but it still allows to add `compatible-with` even when subtype `vnd.elasticsearch` is not present. The version won't be parsed though

pgomulka commented:
I have reworked the regex. I tried not to be overly strict as there has to be some flexibility to allow additional subtypes defined in SQL plugin (see `TextFormat` enum)
the regex now allows for `compatible-with` parameter only to be present when `vnd.elasticsearch` subtype is used.

pgomulka committed:
code review follow up

pgomulka committed:
unused imports

pgomulka committed:
revert unrelated test

pgomulka commented:
I think that if we want to make the `media-type` parsing more strict, we have to refactor the `XContentType.fromMediaType` and `XContentType.fromMediaTypeOrFormat`
I think this methods should be refactored to either parse mediaType (application/json) or a format (a parameter like `json`, `csv`) 
I think it is at the moment not clear what method to use and there is a lot of usages of  `XContentType.fromMediaTypeOrFormat` where just `XContentType.fromMediaType` should be used.

I made an attempt here, but I have some SQL test failing now. https://github.com/elastic/elasticsearch/pull/61845

jaymode reviewed (commented):
I left some questions and I also wonder if we really need to use a regex for this parsing? Also, `compatible-with` is a vendor specific parameter if I understand correctly but I am not sure that there should be a requirement that this parameter comes first vs charset being sent in the header first?

pgomulka commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -130,7 +154,7 @@ public static XContentType fromMediaTypeOrFormat(String mediaType) {
         }
         final String lowercaseMediaType = mediaType.toLowerCase(Locale.ROOT);
-        if (lowercaseMediaType.startsWith("application/*")) {
+        if (lowercaseMediaType.startsWith("application/*") || lowercaseMediaType.equals("*/*")) {
```
I incorrectly read the SQL test, will remove

pgomulka commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -114,13 +116,35 @@ public XContent xContent() {
+            // custom subtype and a version: vnd.elasticsearch+json;compatible-with=7
+            "((vnd\\.elasticsearch\\+([^;\\s]+)(\\s*;\\s*compatible-with=(\\d+)))" +
+            "|([^;\\s]+))" + //subtype: json,yaml,etc some of these are defined in x-pack so can't be enumerated
+            "(?:\\s*;\\s*(charset=UTF-8)?)?$",
```
good point. What is more there could other parameters which we do not expect from server perspective.
SQL for instance has `header=absent` parameter 
So this should either be any text or a sequence of `; key=value`

pgomulka commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -157,6 +183,30 @@ public static XContentType fromMediaType(String mediaType) {
         return null;
     }
 
+    //public scope needed for text formats hack
```
this possibly was a stale comment - removed and changed to private

pgomulka committed:
remove */* and method visibility

pgomulka commented:
> I left some questions and I also wonder if we really need to use a regex for this parsing? 

Do you think something as simple as (`contains("vnd.elasticsearch")` or similar would be enough? It would be harder to perform validation of the format
 

> Also, compatible-with is a vendor specific parameter if I understand correctly but I am not sure that there should be a requirement that this parameter comes first vs charset being sent in the header first?

You are right, I think there is no need for `compatible-with` to be present first on the parameter list. charset, header or similar parameters can be first (I think the order can be random) - will fix

jaymode commented:
> Do you think something as simple as (contains("vnd.elasticsearch") or similar would be enough? It would be harder to perform validation of the format

I was thinking about the possibility of having a new class for doing the parsing to move it away from `XContentType` and `TextFormat` in SQL. 

```java
public class ContentTypeParser {

    public ContentTypeParser(Set<String> acceptedMimeTypes) { // plugins should be able to provide these so we don't break sql. maybe need more than a string?

    }

   public ContentType parse(String contentTypeHeader) {
      // split string on semicolon
      // validate media type is accepted (IIRC whitespace is ok so trim it)
      // rest of strings are params. validate per RFC 7230 and use ones that we care about
     // or use a regex and we can change if necessary
   }

  public ContentType fromAcceptHeader(String acceptHeader) { // not necessary for this PR but I guess it is for SQL

   }
}

public class ContentType {
    public String mediaType() {}
    public Version compatibleWith() {} // or return the int version
    public Charset charset() {} // not necessary for this
}
```

Then `RestRequest` would need to change to use this parser and enable retrieval of the ContentType that was parsed from the request. What do you think of this?

tlrx commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -114,13 +116,35 @@ public XContent xContent() {
 
+    /**
+     * A regexp to allow parsing media types. It covers two use cases.
+     * 1. Media type with a version - requires a custom vnd.elasticserach subtype and a compatible-with parameter
```
`elasticserach` broke my eyes

pgomulka committed:
typos

pgomulka committed:
Merge branch 'master' into compat/content-type-header-with-version

pgomulka committed:
exted parsing with parameters

pgomulka committed:
import

pgomulka committed:
fix test

pgomulka committed:
scope parser

pgomulka committed:
x-ndjson to be versioned

pgomulka commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -112,10 +113,25 @@ public XContent xContent() {
+        .withMediaTypeAndParams("application/vnd.elasticsearch+yaml", YAML, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+cbor", CBOR, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .withMediaTypeAndParams("application/vnd.elasticsearch+x-ndjson", JSON, Map.of("compatible-with", "\\d+","charset", "UTF-8"))
+        .build();
```
defining versioned media types up front, means that we won't throw a exception if someone specifies it with oss licence
if he uses oss on v8 server for api that was removed and provides `application/vnd.elasticsearch+json;compatible-with=7` he will get a 404

is there an easy way to make `XContentType` plugin aware? or distro aware?

pgomulka replied:
but this also would affect plugins like SQL. In oss we don't know about MediaTypes which are defined over there
if we throw an exception for a unknown type I am worried we might end up with a logic based on exceptions
```
try XContentType.parseMediaType
try XContentType.parseFormat
try TextFormat.parseMediaType
try TextFormat.parseFormat
```

@bpintea any thoughts on this? I guess SQL had to face this in the past

jakelandis replied:
Is there any way to register a custom parser ? or just add the SQL values to the core parser and specificy which rule set to use when parsing ?

pgomulka replied:
that is an interesting idea.. We could have something like a SqlMediaTypeParser and it would either return TextFormat or XContentType. The parsing logic would remain the same in the MediaTypeParser. The idea of SqlMediaTypeParser woudl be just to abstract on the setup of the parser.
The same could be done for CompatibleApiMediaTypeParser and we could keep it in xpack plugin.
this does not solve the problem of trying to parse a format, then if failed to parse a media type. I think we should try to just check for presence of the format parameter and if present to used. Otherwise trying to use Accept Header.
I think also that Content-Type should not be used for response formatting in SQL plugin. It looks like it is allowed in code, but I don't think it would work (it would fail when parsing that header in server)

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -65,8 +51,10 @@ public T fromFormat(String format) {
+     *
      * @param headerValue a header value from Accept or Content-Type
      * @return a parsed media-type
+     * //todo pg should this all be maybe based on a regex?
```
I suggest that we benchmark this with a JMH microbenchmark (separately) with a variety of different content types and parameters

pgomulka replied:
I wrote a benchmark and regex approach is slower. I does cover more (the exact characters allowed by rfc) but not sure we need this. See the implementation:
https://github.com/pgomulka/elasticsearch/blob/compat/benchmark_regex_vs_splits/benchmarks/src/main/java/org/elasticsearch/benchmark/MediaTypeParsingBenchmark.java

https://github.com/pgomulka/elasticsearch/blob/compat/benchmark_regex_vs_splits/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParserRegex.java

```
Benchmark                                     Mode  Cnt  Score   Error  Units
MediaTypeParsingBenchmark.parseIter           avgt   30  0.801 ± 0.010  us/op
MediaTypeParsingBenchmark.parseIterNoParams   avgt   30  0.200 ± 0.002  us/op
MediaTypeParsingBenchmark.parseRegex          avgt   30  1.639 ± 0.012  us/op
MediaTypeParsingBenchmark.parseRegexNoParams  avgt   30  0.504 ± 0.011  us/op

```

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -96,6 +91,17 @@ public ParsedMediaType parseMediaType(String headerValue) {
     }
 
+    private boolean isValidParameter(String typeWithSubtype, String parameterName, String parameterValue) {
+        if(parametersMap.containsKey(typeWithSubtype)){
```
```suggestion
        if (parametersMap.containsKey(typeWithSubtype)) {
```

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -96,6 +91,17 @@ public ParsedMediaType parseMediaType(String headerValue) {
+    private boolean isValidParameter(String typeWithSubtype, String parameterName, String parameterValue) {
+        if(parametersMap.containsKey(typeWithSubtype)){
+            Map<String, String> parameters = parametersMap.get(typeWithSubtype);
+            if(parameters.containsKey(parameterName)){
```
```suggestion
            if (parameters.containsKey(parameterName)) {
```

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -96,6 +91,17 @@ public ParsedMediaType parseMediaType(String headerValue) {
+            Map<String, String> parameters = parametersMap.get(typeWithSubtype);
+            if(parameters.containsKey(parameterName)){
+                String regex = parameters.get(parameterName);
+                return parameterValue.matches(regex);//todo pg should we precompile regex?
```
We should compile the regex into a pattern to avoid doing it everytime

pgomulka replied:
will do

jaymode commented on libs/x-content/src/test/java/org/elasticsearch/common/xcontent/MediaTypeParserTests.java:
```diff
@@ -29,37 +29,42 @@
+
+    MediaTypeParser<XContentType> mediaTypeParser = new MediaTypeParser.Builder<XContentType>()
+        .withMediaTypeAndParams("application/vnd.elasticsearch+json",
+            XContentType.JSON,Map.of("compatible-with","\\d+",
+            "charset","UTF-8"))
```
```suggestion
            XContentType.JSON, Map.of("compatible-with", "\\d+",
            "charset", "UTF-8"))
```

pgomulka committed:
Merge branch 'master' into compat/content-type-header-with-version

jaymode commented on server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java:
```diff
@@ -93,4 +94,66 @@ public void testFromRubbish() throws Exception {
+    }
+
+    public void testVersionParsing() {
+        byte version = (byte)Math.abs(randomByte());
```
```suggestion
        byte version = (byte) Math.abs(randomByte());
```

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -86,7 +76,12 @@ public ParsedMediaType parseMediaType(String headerValue) {
-                        parameters.put(keyValueParam[0].toLowerCase(Locale.ROOT), keyValueParam[1].toLowerCase(Locale.ROOT));
+                        String parameterName = keyValueParam[0].toLowerCase(Locale.ROOT);
+                        String parameterValue = keyValueParam[1].toLowerCase(Locale.ROOT);
+                        if(isValidParameter(typeWithSubtype,parameterName,parameterValue) == false) {
```
```suggestion
                        if (isValidParameter(typeWithSubtype, parameterName, parameterValue) == false) {
```

jaymode commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -86,7 +76,12 @@ public ParsedMediaType parseMediaType(String headerValue) {
+                        String parameterName = keyValueParam[0].toLowerCase(Locale.ROOT);
+                        String parameterValue = keyValueParam[1].toLowerCase(Locale.ROOT);
+                        if(isValidParameter(typeWithSubtype,parameterName,parameterValue) == false) {
+                            return null;
```
should we throw an exception here? Maybe one that returns a 400 since the parameter is invalid

pgomulka replied:
but should it only be here? or should we throw exceptions in other places where we return null.
I think we should throw an exception when a media type does not conform to rfc
but if we don't know about the requested media type I think it might cause problems in code at the moment (we rely on checking if parsing returned null and then try a different method. See https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java#L100

pgomulka replied:
as mentioned here this PR will not introduce throwing exceptions yet. I will follow up in another PR

jakelandis replied:
Can you log an issue for this ?

pgomulka replied:
done - https://github.com/elastic/elasticsearch/issues/63080 will try to gather more details before starting work on it

jaymode commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java:
```diff
@@ -294,7 +295,11 @@ public String subtype() {
 
-    private static final MediaTypeParser<TextFormat> parser = new MediaTypeParser<>(TextFormat.values());
+    private static final MediaTypeParser<TextFormat> parser = new MediaTypeParser.Builder<TextFormat>()
+        .withMediaTypeAndParams(PLAIN_TEXT.typeWithSubtype(), PLAIN_TEXT, Map.of("header", "present|absent", "charset", "utf-8"))
```
I am concerned that having charset with a regex of utf-8 could break someone using sql. This should be checked with the sql team

pgomulka replied:
this only means that a parameter is possible. if it won't be provided it won't fail any validation.
I asked SQL team about this offline.
 @bpintea  can you confirm that we can expect any of these in SQL? Do you expect any other charsets?

jaymode replied:
Not speaking for SQL, but in general with a lot of clients on Windows, I have seen them send `charset=ISO-8859-1` with the `Content-Type` header and would operate under the assumption that they may also send the charset with the `Accept` header; nothing in RFC 7231 would prevent that. Additionally, multiple accept headers are allowed so you can wind up with something like:

```
Accept: text/plain;charset="utf-8", text/plain;charset="ISO-8859-1"
```

There is also the `Accept-Charset` header which technically should have precedence over the media type parameter in the `Accept` header. (It is kind of a mess)

pgomulka replied:
I think having multiple headers would be beyond the scope of this change. We would also have to support the `q` factor to understand the preference of clients Accept type.

bpintea replied:
> if it won't be provided it won't fail any validation.

Right, I believe currently the `TextFormat`er simply ignores an `Accept`-requested charset (that happens to be different than UTF-8). Hasn't been an issue so far, but it might make sense to do something about it at some point.

pgomulka committed:
Apply suggestions from code review

Co-authored-by: Jay Modi <jaymode@users.noreply.github.com>

pgomulka committed:
precompile pattern

pgomulka committed:
import

jakelandis reviewed (commented):
a few comments

pgomulka commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -120,4 +125,36 @@ public T getMediaType() {
+            return this;
+        }
+
+        public Builder<T> withMediaTypeAndParams(String alternativeMediaType, T mediaType, Map<String, String> paramNameAndValueRegex) {
```
I will have to copy the pattern and create a new one to make sure  case insensitivity is used. I will give a go and see how it works

pgomulka committed:
removal of unused method and pattern in method

pgomulka committed:
throwing exception when failing to parse

checkstyle

test fixes

*/* header

catch

hacks around text/plain

sout remove

precommit

fix nullpointer

fix yml test

fix yml test

pgomulka committed:
throwing exception when failing to parse

pgomulka committed:
checkstyle

pgomulka committed:
test fixes

pgomulka committed:
*/* header

pgomulka committed:
catch

pgomulka committed:
hacks around text/plain

pgomulka committed:
sout remove

pgomulka committed:
precommit

pgomulka committed:
fix nullpointer

pgomulka committed:
fix yml test

pgomulka committed:
fix yml test

pgomulka commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -1900,9 +1900,9 @@ protected final ElasticsearchStatusException parseResponseException(ResponseExce
-        if (xContentType == null) {
-            throw new IllegalStateException("Unsupported Content-Type: " + entity.getContentType().getValue());
-        }
+//        if (xContentType == null) {
```
an exception is thrown when parsing

pgomulka replied:
not throwin exception in this PR

pgomulka commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -107,16 +61,16 @@ public RestResponse buildResponse(SqlQueryResponse response) throws Exception {
                 // XContent branch
-                if (xContentType != null) {
-                    XContentBuilder builder = channel.newBuilder(request.getXContentType(), xContentType, true);
+                if (responseMediaType != null && responseMediaType instanceof XContentType) {
```
when parsing returns a supertype to `TextFormat` and `XContentType` it makes it harder to implement these two branches.
i think we should probably move the formatting behaviour out of TextFormat. It should only implement methods defined on MediaType (the original XContentType methods)

pgomulka replied:
leaving this up to a discussion. With this approach we can have a single parser that returns both XContentType and TextFormat under MediaType interface
XContentType does not implement format method. But maybe these two should have a common createRestResponse method?

pgomulka commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java:
```diff
@@ -0,0 +1,77 @@
+
+        //todo pg we should probably remove this possibility. it should only be accept or format.
+        // also if you specify sql's textformat in content type it will fail with exception now
+        String contentType = request.header("Content-Type");
```
I made a comment to suggest we get rid of using Content-Type to set responseType. However I think most sql queries contain a body, so using a Content-Type to default responseType might be ok.
If however a request did not contain a body, then it would not make sense. I don't know we have any SQL api though.
However, this obviously means that only json/cbor/smile/yml are allowed .

pgomulka replied:
also not fixing this in current PR. might be worth discussing with @elastic/es-ql

bpintea replied:
> I think most sql queries contain a body

Indeed, all requests concerning this code always contain a body.
(In the SQL API only one API lacks it, the one for stats inquiring, which won't hit this code tho.)

pgomulka commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestResponse.java:
```diff
@@ -45,15 +45,21 @@
-            this.bodyContentType = XContentType.fromMediaType(contentType);
+            // todo pg it feels like there is a lot of 'ifs' around text, csv etc being returned. shouldn't we treat text as the same json and others?
+            // we are testing sql formats with this client, but don't have access to sql parsing code
+            try{
```
this client does not have access to SQL code, but has to support these formats too.

pgomulka replied:
we are not throwing an exception when parsing in this PR

pgomulka commented on server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java:
```diff
@@ -98,10 +101,17 @@ public XContentBuilder newBuilder(@Nullable XContentType requestContentType, boo
-                responseContentType = XContentType.fromMediaType(acceptHeader);
+            if (Strings.hasText(format)) {
+                responseContentType = XContentType.fromFormat(format);
+            } else if (Strings.hasText(acceptHeader) && acceptHeader.equals("*/*") == false
```
since parsing is strict and throws an exception when unkown type is used then it requires a lot of "if guards" to make sure it will not fail.
Also looks to me that `text/plain` is commonly used as returnType for `_cat` api. We should probably add it to xcontentType?

pgomulka replied:
not fixing this in current PR. will follow up

pgomulka commented:
@jakelandis @jaymode from previous comment the conclusion was to make parsing strict, throwing exception when unsupported type, parameters or format is wrong.
I like the idea, but I think this requires much more work then the scope of this PR. There is a lot of usages of `defautling to null` logic around content-type and accept header parsing.
I left comments and `//todo pg` to start discussions. 
My suggestion would be that we don't implement strict parsing in this PR. We would still return null as we used to and only focus on adding parsing `compatible-with` and other parameters.
We can follow up with another PR to make the parsing strict.
Let me know what you think.

pgomulka commented on server/src/main/java/org/elasticsearch/rest/BytesRestResponse.java:
```diff
@@ -83,6 +83,7 @@ public BytesRestResponse(RestStatus status, String contentType, byte[] content)
      * Creates a binary response.
      */
+    //todo pg this should probably be acceptType? (2nd param)
     public BytesRestResponse(RestStatus status, String contentType, BytesReference content) {
```
should this be renamed to acceptType?

pgomulka replied:
renamed.

pgomulka commented on buildSrc/src/main/java/org/elasticsearch/gradle/http/WaitForHttpResource.java:
```diff
@@ -145,6 +145,7 @@ HttpURLConnection buildConnection(SSLContext ssl) throws IOException {
         configureSslContext(connection, ssl);
         configureBasicAuth(connection);
         connection.setRequestMethod("GET");
+        connection.setRequestProperty("Accept", "text/plain; charset=UTF-8");
```
this prevents defaulting Accept header to `text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2` by `HttpUrlConnection`

pgomulka committed:
Revert "throwing exception when failing to parse"

This reverts commit bd4dee1126974d0cf66d63eaf62564b8c3d9f032.

jakelandis commented:
> We can follow up with another PR to make the parsing strict.

That sounds good. Let's keep this PRs scope under control. 

> I left comments and //todo pg to start discussions.

Can you avoid merging these todo's to master.  If there is something concrete a future developer can understand and address, that is fine for a TODO otherwise please log an issue and make notes there, or just keep personal notes to follow up.

pgomulka commented:
> Can you avoid merging these todo's to master. If there is something concrete a future developer can understand and address, that is fine for a TODO otherwise please log an issue and make notes there, or just keep personal notes to follow up.

these will all be removed. I just left them along the development to not forget to make comments. And then didn't want to remove immediately so it won't get invalidated to early. Will try to make comments on a PR only

jakelandis commented:
@pgomulka - I am a bit lost in the state of the review with the simplified scope (We can follow up with another PR to make the parsing strict.) ... Is it possible to resolve any conversations which are now out of scope and get all of the tests to pass ?

pgomulka committed:
Merge branch 'revert_strict_header_parsing' into compat/content-type-header-with-version

pgomulka committed:
import

pgomulka committed:
fix storing response time in sql

pgomulka committed:
return null on errors

pgomulka committed:
sqlmedia type parsing test - do not throw exceptions

pgomulka committed:
remove todoes

pgomulka committed:
rename argument

jakelandis commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java:
```diff
@@ -113,9 +115,26 @@ public XContent xContent() {
+        .withMediaTypeAndParams("application/smile", SMILE, Collections.emptyMap())
+        .withMediaTypeAndParams("application/cbor", CBOR, Collections.emptyMap())
+        .withMediaTypeAndParams("application/json", JSON, Map.of("charset", Pattern.compile("UTF-8")))
+        .withMediaTypeAndParams("application/yaml", YAML, Map.of("charset", Pattern.compile("UTF-8")))
```
I think this is technically non-passive (only supporting UTF-8 encoding) and should probably be mentioned in the breaking changes ... thus only targeting 8.0 and (outside of this PR) ensure that a 7.x server maps the compatible header to standard header ... i.e. (application/vnd.elasticsearch+yaml; compatible-with=7 == application/yaml)

pgomulka replied:
that's a good point. I think clients will have to upgrade to latest 7.x client to get `compatible-with`  sent on their request.
 We should also support that in 7.x server. Will raise an issue/pr

pgomulka replied:
handling of versioned media types in a form `application/vnd.elasticsearch+json;compatible-with=7` will be handled in this pr https://github.com/elastic/elasticsearch/pull/63071

jakelandis commented on server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java:
```diff
@@ -23,6 +23,7 @@
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.nullValue;
 
 public class XContentTypeTests extends ESTestCase {
```
Can you also add a small test for (lowercase charset=utf-8)

jakelandis commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java:
```diff
@@ -120,4 +125,37 @@ public T getMediaType() {
+            for (Map.Entry<String, Pattern> params : paramNameAndValueRegex.entrySet()) {
+                String parameterName = params.getKey().toLowerCase(Locale.ROOT);
+                Pattern parameterRegex = params.getValue();
+                Pattern pattern = Pattern.compile(parameterRegex.pattern(), Pattern.CASE_INSENSITIVE);
```
can you remove the insensitive recompile here and and just use the pattern as-is (and move the insenstivie up to the declration)

pgomulka replied:
I will apply `toLowerCase` for keys though

jakelandis replied:
OK ... in that case, maybe go back to a `Map<String, String>` as the collection passed ?  It is a bit confusing to compile it, then re-compile it.  Sorry, when i first suggested to pass `Map<String, Pattern>` I had envisioned only compiling once.

pgomulka replied:
agree, will refactor this back

jakelandis commented on server/src/test/java/org/elasticsearch/common/xcontent/XContentTypeTests.java:
```diff
@@ -93,4 +94,65 @@ public void testFromRubbish() throws Exception {
+
+    public void testVersionedMediaType() {
+        String version = String.valueOf(Math.abs(randomByte()));
+        assertThat(XContentType.fromMediaType("application/vnd.elasticsearch+json;compatible-with=" + version),
```
can you also add a test that has both compatible-with and charset ?

pgomulka replied:
this specific test is only meant to test versioning. I covered multiple parameters support in `MediaTypeParserTests.testJsonWithParameters`

jakelandis replied:
thanks... i missed that!

jakelandis reviewed (commented):
Changes look good. A few minor comments. However, could we get someone from @elastic/es-ql to take a look a the the changes for SQL ?

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

jakelandis commented:
I updated the label to be only for 8.0 and >breaking

For clarity, the breaking here is that with this change we will only support requests for charset=UTF-8. Prior to this change we accepted any charset request...but _I think_ always encoded in UTF-8. So it is non-passive in the fact we are more strict about what we accept. With this PR we will not throw an exception if we don't support it (to avoid scope creep), however if the charset (or any parameters) are not supported it will result in the default JSON return type. "Accept: application/yaml;charset=junk" results in application/json.  A follow up PR will throw the necessary exception to result in the more correct 40x error.

pgomulka committed:
review follow up

pgomulka committed:
remove pattern from signature

pgomulka committed:
fix regex

jakelandis reviewed (approved):
LGTM, pending someone from @elastic/es-ql validating the SQL part(s).

jaymode commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -107,16 +61,16 @@ public RestResponse buildResponse(SqlQueryResponse response) throws Exception {
-                if (xContentType != null) {
-                    XContentBuilder builder = channel.newBuilder(request.getXContentType(), xContentType, true);
+                if (responseMediaType != null && responseMediaType instanceof XContentType) {
+                    XContentType type = (XContentType)responseMediaType;
```
nit:

```suggestion
                    XContentType type = (XContentType) responseMediaType;
```

jaymode reviewed (approved):
LGTM

pgomulka committed:
Apply suggestions from code review

Co-authored-by: Jay Modi <jaymode@users.noreply.github.com>

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java:
```diff
@@ -0,0 +1,75 @@
+            Map.of("header", "present|absent", "charset", "utf-8"))
+        .withMediaTypeAndParams(TextFormat.CSV.typeWithSubtype(), TextFormat.CSV,
+            Map.of("header", "present|absent", "charset", "utf-8",
+                "delimiter", "[^\"\n\r\t]+"))
```
Is the delimiter regex derived from [TextFormat.CSV#delimiter()](https://github.com/elastic/elasticsearch/blob/306e1d76da3db652db57f11f847705b3995609ff/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/TextFormat.java#L152)? Because it specifies what characters are allowed, inline with what that method checks, on one hand, but allows a count of multiple character, out of line with a check there, on the other.
The exception messages in the method are explanatory and I'd ever so slightly prefer to return those (and maybe don't enforce a specific set here), but if you'd still like to enforce the param format in the parser definition, I'd suggest to maybe also limit the char count to one.

pgomulka replied:
good point, I misread that method. It should only be one character.
The exception messages are indeed explanatory in this method. With the parameter validation when parsing, we will loose them - that code will not be reachable as it will fail earlier when parsing

pgomulka replied:
I changed this to expect the presence of a `delimiter` parameter, but the pattern will be `.+` which will always pass, and the validation will be done in `TextFormat.CSV#delimiter()`

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java:
```diff
@@ -0,0 +1,75 @@
+            && (sqlRequest.binaryCommunication() == null || sqlRequest.binaryCommunication())) {
+            // enforce CBOR response for drivers and CLI (unless instructed differently through the config param)
+            return XContentType.CBOR;
+        } else if (request.hasParam(URL_PARAM_FORMAT)){
+            return validateColumnarRequest(sqlRequest.columnar(), parser.fromFormat( request.param(URL_PARAM_FORMAT)));
```
nit, spacing
```suggestion
        } else if (request.hasParam(URL_PARAM_FORMAT)) {
            return validateColumnarRequest(sqlRequest.columnar(), parser.fromFormat(request.param(URL_PARAM_FORMAT)));
```

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java:
```diff
@@ -0,0 +1,75 @@
+        return validateColumnarRequest(sqlRequest.columnar(), parser.fromMediaType(contentType));
+    }
+
+    private MediaType validateColumnarRequest(boolean requestIsColumnar, MediaType fromMediaType) {
```
optional, could be made `static`.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParserTests.java:
```diff
@@ -0,0 +1,90 @@
+        assertThat(text, is(TSV));
+    }
+
+    public void testParametersParsing() {
```
Since the methods only tests media type correct detection - i.e. the other parameters in the `Accept` aren't analysed by `SqlMediaTypeParser`, nor tested here, right? -, I would maybe propose to rename it (`testMediaTypeDetection` or similar?).

(Slightly off-topic, the `header` parameter will need a rework or removal, as per #60271, but that's not the object of this PR).

bpintea reviewed (approved):
The SQL bits look good, I've left only some minor comments.

pgomulka committed:
Apply suggestions from code review

Co-authored-by: Bogdan Pintea <bpintea@gmail.com>

pgomulka committed:
relax delimiter validation
